### DISCUSSION TITLE
DSI LCD (like waveshare clone) and lysmarine

### DISCUSSION THREAD

**Discussion by atchik6056** [bareboat-necessities/lysmarine_gen#327] [Q&A] [answered]
Hello
I connected DSI and HDMI screens to PI4. Everything works ok with HDMI or VNC but seems that BBN application autostart kills DSI touchscreen. It works but completely misbehaves  If I comment autostart out ( $HOME/.config/openbox/autostart.desktop) then all apps are available via budgie menu and all works ok without BBN desktop.
Is it expected? Any comment?
PS: lysmarine is 20230702 version
thank you
A

**Comment by mgrouch** [comment]
If you want to connect two screens you are on your own. You will need to figure it out for your own particular hardware set up. 

BBN default setup works only for one screen

**Comment by atchik6056** [comment]
I am not talking about 2 screens, but rather compare functionality with and without desktop apps autostart. If it's enabled, HDMI or VNC works, reacting correctly on USB connected keyboard and mouse, starting apps and etc, but if the only DSI touchsceen connected, then it just pops up budgie menu and onboard keyboard on whatever touch on the touch screen. It looks like calibration is completely off. However everything functions normal if I just remove BBN apps autostart from $HOME/.config for onbox. Then budgie session and onbox WM works perfectly fine., supporting 2-3 finger zoom, except opencpn though. it seems that BBN autostart ruins the touchscreen. Maybe silly question but what if I keep it commented out , and just add big application icons on the standard budgie/gnome/xfce desktop?

**Comment by atchik6056** [comment]
if only HDMI attachment is working, and recommended, then I can use HDMI switch to switch between cockpit and cabin, similar to one I use at home which switches 2 comps to the one screen/USB dongle mouse/key and cost $20/30 at aliexpress.

**Comment by mgrouch** [comment]
As far as I know it works with DSI screens (at least some). But you need to set it up with DSI screen connected. If you setup with HDMI screen then plug DSI and boot with it you might have weird behavior. 

autostart.desktop calls autostart script. 
you can comment out lines calling 
evdev-rce and twofing

**Comment by mgrouch** [comment]
Switching between DSI and HDMI at run-time (or running both) would cause issues with touch calibration.
As well as having two HDMI screens with different resolutions

**Comment by atchik6056** [comment]
Hi Michael
I did another test by reinstalling lysmarine and having pi connected to router via cable. No HDMI, only touchscreen.. After driver is installed I got the same: as with HDMI connected install: desktop of lysmarine shown up ok, but any touch causes the only gnome menu and onboard key popup. Icons do not work as static picture, similar as with attached HDMI installation before. Calibration is completely off. If I comment onbox auto-start as I mention before, then all apps work ok from gnome menu. So something in the desktop auto start causing weird touch calibration.  Just as FYI. DSI LCD touch screen is recommended by openplotter as default https://www.waveshare.com/7inch-dsi-lcd.htm.
thanks
Alex

**Comment by mgrouch** [comment]
Which BBN image version?

Also your screen is 800x480 resolution which is a bit low for OpenCPN. You will have issues with dialog boxes not fitting on screen

**Comment by mgrouch** [comment]
Also have you tried these steps?

https://bareboat-necessities.github.io/my-bareboat/bareboat-os.html#_touchscreen

**Reply by atchik6056** [reply]
yes I did these steps, it didn't change anything, so status now:
1) touchscreen does react on any touches with poping/minimizing gnome/budgie menu and onboard, though it shows all the icons correctly, but they are just static. The BBN desktop shows on the touchscreen, but not on HDMI if attached, while booting process can be seen on both.
2) the only way to control BBN is to attach USB mouse and keyboard, then everything works ok and all icons are clickable. 
3) if I comment out BBN auto-start all together or just comment out in thethe following lines  as you mentioned:
```
(openbox --replace) # &
#(evdev-rce || true) &
#(twofing --click=first --grab-by-disabling /dev/twofingtouch || true) &
```
then I can launch any BBN apps from the budgie menu on the touchscreen and everything seems ok.
3) BBN is 20230702
4) xinput_calibrator shows only one point somewhere in the muddle of the screen (supposed to be 4) and exit without any errors

**Reply by atchik6056** [reply]
on (3) on above: touchscreen works ok including opencpn but not 2-3 finger /zoom

**Comment by mgrouch** [comment]
Check your /usr/share/X11/xorg.conf.d/90-touchinput.conf file:

````
cat /usr/share/X11/xorg.conf.d/90-touchinput.conf
````

What is the output for you? What does MatchProduct say?

If MatchProduct string in it is empty you will need to put it in manually. 

First find the product:

````
cat /var/log/Xorg.0.log | grep TOUCH
````

Did you find your product name?

Then copy and paste product name into MatchProduct section

````
sudo nano /usr/share/X11/xorg.conf.d/90-touchinput.conf
````

and save the file.

Also due to other fixes to USB I suggest you
to install 2023-07-23. 


Do you even have twofing link of /dev filesystem?

````
ls  -l  /dev/twofing
````

There is 

https://raw.githubusercontent.com/Plippo/twofing/master/rules/70-touchscreen-waveshare.rules

udev rule for waveshare screen which you might need to edit to match your product and put into /etc/udev/rules.d/


````
sudo curl -o /etc/udev/rules.d/70-touchscreen-waveshare.rules  https://raw.githubusercontent.com/Plippo/twofing/master/rules/70-touchscreen-waveshare.rules

sudo nano /etc/udev/rules.d/70-touchscreen-waveshare.rules  

````

**Reply by atchik6056** [reply]
thank you for the hints, I ll try it ASAP
yes, File /usr/share/X11/xorg.conf.d/90-touchinput.conf 
```
Section "InputClass"
    Identifier "calibration"
    Driver "evdev"
    MatchProduct "WS Capacitive TouchScreen WS_xinchDSI_Touch_c bcm2835 (i2c@7e804000)"
```
FYI: There is a bug in waveshare installer, I am in contact with them, but for now I fixed it by changing /usr/share/X11/xorg.conf.d/90-touchinput.conf manually as above.
Off my curiosity and to check hw i flashed cd with openplotter and after above workaround and tons of upgrades, everything works, touchscreen and HDMI,  even if I plug it in and out (though touch might be off calibrated)

**Reply by mgrouch** [reply]
Does multitouch work for your in OpenCPN under Openplotter? The only way in current state of the things it can work there is when you install OpenCPN flatpak (which delivers you other issues with VNC) under openplotter. 

It installs OpenCPN with (originally my patch, now derivative of it ) for touchscreen with wxWidgets 3.2. Here is the story 

https://github.com/OpenCPN/OpenCPN/pull/2333

(Took years to integrate due to wxWidgets moving really slow)

wxWidgets 3.2 is not something supported by Debian under Bullseye still. Another is issue is that running wxWidgets 3.2 and 3.0 takes more resources (memory). So I wouldn’t recommend using flatpak OpenCPN on bullseye. Flatpak is poor man solution for running foreign more cuttings age things on otherwise tested and consistent operating system.

**Comment by mgrouch** [comment]
Looks like that DSI Waveshare screen requires a kernel module. The source code is not published. And it’s not built for all kernels Raspberry Pi OS can come with. 

Was it a good choice to pick that screen model?

**Reply by atchik6056** [reply]
you are correct, I overlooked it, I think it is not the best choice at least for now

**Comment by atchik6056** [comment]
I managed to make it working using  2023-07-23. No HDMI initially. All was good after I installed touch driver and commented out autostart as :
```
 (openbox --replace) # &
 #(evdev-rce || true) &
 #(twofing --click=first --grab-by-disabling /dev/twofingtouch || true) &
```
also I modified /etc/udev/rules.d/70-touchscreen-waveshare.rules as you mentioned above to my device "WS Capacitive TouchScreen WS_xinchDSI_Touch_c bcm2835 (i2c@7e804000)"
Touchscreen works starting all the apps, dragging maps and so on....except twofing  and right click obviously.  BBN dashboard shows all apps and behave as expected. xinput_calibrator works fine. /dev/twofingtouch exists:
```
user@lysmarine:~ $ ls -al /dev/two*
lrwxrwxrwx 1 root root 12 Aug  2 02:27 /dev/twofingtouch -> input/event2
```
what do you think about: https://www.embeddedtouchpanelpc.com/china-full_ip65_waterproof_lcd_display_resistive_touch_display_7_inch_size-10551098.html

**Comment by mgrouch** [comment]
Regarding Waveshare DSI touchscreen issues

here is a thread on raspberry pi forums:

https://forums.raspberrypi.com/viewtopic.php?t=366011

**Reply by atchik6056** [reply, marked as answer]
I have contacted Waveshare last year when I was having the problem, I think they fixed installation script and it works ok.
The following configuration works for me very good, some I mentioned above:
1) Run installation script provided by Waveshare https://github.com/waveshare/Waveshare-DSI-LCD
2) Make sure you have something like this in /boot/config.txt (supposed to be added by installation script)

````

## https://www.waveshare.com/wiki/7inch_HDMI_LCD_(C)
max_usb_current=1
#hdmi_group=2
#hdmi_mode=87
#hdmi_cvt 800 480 60 6 0 0 0
#hdmi_cvt 1024 600 60 6 0 0 0
#hdmi_cvt 1280 800 60 6 0 0 0
#hdmi_drive=1
ignore_lcd=1
dtoverlay=vc4-kms-v3d
dtparam=i2c_vc=on
dtoverlay=WS_xinchDSI_Screen,SCREEN_type=2,I2C_bus=1
dtoverlay=WS_xinchDSI_Touch,I2C_bus=1

````

3) Make sure you have something like this in /usr/share/X11/xorg.conf.d/90-touchinput.conf

````

Section "InputClass"
    Identifier "calibration"
    Driver "evdev"
    MatchProduct "WS Capacitive TouchScreen WS_xinchDSI_Touch_cbcm2835 (i2c@7e804000)"
    MatchDevicePath "/dev/input/event*"
    Option "EmulateThirdButton" "1"
    Option "EmulateThirdButtonTimeout" "750"
    Option "EmulateThirdButtonMoveThreshold" "30"
EndSection

````

4) From time to time i need to run calibration which updates /etc/X11/xorg.conf.d/99-calibration.conf

````

Section "InputClass"
        Identifier      "calibration"
        MatchProduct    "WS Capacitive TouchScreen"
        Option  "MinX"  "-427"
        Option  "MaxX"  "65962"
        Option  "MinY"  "2330"
        Option  "MaxY"  "64516"
        Option  "SwapXY"        "0" # unless it was already set to 1
        Option  "InvertX"       "0"  # unless it was already set
        Option  "InvertY"       "0"  # unless it was already set
EndSection

````

thank you

**Reply by mgrouch** [reply]
Thanks. It’s good to know

**Comment by mgrouch** [comment]
Similar problem resolved for another DSI attached screen:

See: https://github.com/bareboat-necessities/lysmarine_gen/issues/416
